docs/presentation/Hierarchies/GroupingRule.md

```diff
@@ -7,99 +7,282 @@ Grouping rules provide advanced ways to group instances when creating hierarchie
+| `label`     | No        | `string` | `"{from value} - {to value}"`                                       | Grouping node's label. May be [localized](../Advanced/Localization.md). |
+
+Range [`fromValue`, `toValue`] is inclusive on both sides. If a value falls into more than one range, the first listed range that contains the value is chosen.
+
+```ts
+[[include:Hierarchies.Grouping.PropertyGroup.Ranges.Ruleset]]
+```
+
+![Example of using "ranges" attribute](./media/hierarchy-grouping-with-ranges-attribute.png)
+
+#### Deprecated Attribute: `groupingValue`
+
+> **Default value:** `"DisplayLabel"`
+
+Specifies whether instances should be grouped using property's display or raw value.
+
+> **Note:** Grouping by property value is required if the display label is overridden to display grouped instances count.
+>
+> **Warning:** Grouping by label and sorting by property value is not possible.
+
+Display value should always be used for grouping. In cases when there's a need to show grouped instances count suffix, that
+can be achieved at the UI component layer by composing UI node's label from node's display label and [GroupingNodeKey.groupedInstancesCount]($presentation-common).
+
+#### Deprecated Attribute: `sortingValue`
+
+> **Default value:** `"DisplayLabel"`
+
+Specifies whether nodes should be sorted by their display label or the grouping property's value. In most cases the result is the same,
+unless a [label override rule](../Customization/LabelOverride.md) is used to change node's display label.
+
+> **Note:** Sorting by property value only makes sense when instances are grouped by property value as well.
+>
+> **Warning:** Grouping by label and sorting by property value is not possible.
+
+### Same Label Instance Grouping
+
+Allows grouping multiple instances with the same label into one ECInstances type of node. Similar to display label grouping, but instead of showing a grouping
+node with multiple grouped ECInstance nodes, it shows a single ECInstances node which represents multiple ECInstances.
+
+| Name                                              | Required? | Type                       | Default   |
+| ------------------------------------------------- | --------- | -------------------------- | --------- |
+| [`applicationStage`](#attribute-applicationstage) | No        | `"Query" \| "PostProcess"` | `"Query"` |
+
+#### Attribute: `applicationStage`
+
+> **Default value:** `"Query"`
+
+> **Performance note:** Grouping at post-processing step requires loading the whole hierarchy level before
+> returning even the first node - avoid using with large numbers of nodes.
+
+Stage of hierarchy creation at which the rule is applied.
+
+- `Query` groups instances during the query, which can often make use of database indexes and is generally fairly quick. This
+  also means that instances are grouped at the level of a single hierarchy specification and doesn't work when a hierarchy
```

Comment:
It means the grouping doesn't happen under conditions listed in the following sentence. 

Enabling this kind of grouping produces a query like `SELECT DisplayLabel, CONCAT(InstanceKey) FROM (SELECT DisplayLabel, InstanceKey FROM myschema.MyClass) GROUP BY DisplayLabel`. We can't apply query-level grouping under conditions listed in the description. This includes cases, when a hierarchy level is:

- Built from classes without a common base class. We can't apply query-level grouping in this case, because there's no way to define a grouping rule that groups multiple classes (`GroupingRule.class` allows defining only one). Now that I think of it, the same applies for post-processing level grouping, and since it's not even possible to define grouping like this, I think there's no point in mentioning it either.

- Using multiple specifications. This type of grouping is applied at specification level, which means cross-specification grouping is not possible, even if specifications return same type of instances.

- Merged from multiple nested hierarchy levels using the `hideNodesInHierarchy` or `hideExpression` attributes. We may have a hierarchy where parent hierarchy level is hidden, which means the hierarchy level consists of combined child hierarchy levels, e.g.:
  ```
  +-- hidden parent node A
  |   +-- visible child node x
  |   +-- visible child node y
  +-- hidden parent node B
      +-- visible child node x
      +-- visible child node z
  ```
  results in a hierarchy like:
  ```
  +-- visible child node x
  +-- visible child node y
  +-- visible child node x
  +-- visible child node z
  ```
  The two child hierarchy levels are built using different queries, which means they're not affected by query-level grouping. The post-processing level grouping, on the other hand, takes effect **after** we combine them, so it does have effect.

Reply:
Thanks for the detailed explanation.